src/SignalR/server/Core/src/Internal/HubFilterFactory.cs

```diff
@@ -0,0 +1,45 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading.Tasks;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNetCore.SignalR.Internal
+{
+    internal class HubFilterFactory : IHubFilter
+    {
+        private readonly Type _hubFilterType;
+
+        public HubFilterFactory(Type hubFilterType)
+        {
+            _hubFilterType = hubFilterType;
+        }
+
+        public ValueTask<HubResult> InvokeMethodAsync(HubInvocationContext invocationContext, Func<HubInvocationContext, ValueTask<HubResult>> next)
+        {
+            return GetFilter(invocationContext.ServiceProvider, _hubFilterType).InvokeMethodAsync(invocationContext, next);
+        }
+
+        public Task OnConnectedAsync(HubInvocationContext context, Func<HubInvocationContext, Task> next)
+        {
+            return GetFilter(context.ServiceProvider, _hubFilterType).OnConnectedAsync(context, next);
+        }
+
+        public Task OnDisconnectedAsync(HubInvocationContext context, Exception exception, Func<HubInvocationContext, Exception, Task> next)
+        {
+            return GetFilter(context.ServiceProvider, _hubFilterType).OnDisconnectedAsync(context, exception, next);
+        }
+
+        private static IHubFilter GetFilter(IServiceProvider serviceProvider, Type filterType)
+        {
+            var filter = (IHubFilter)serviceProvider.GetService(filterType);
+            if (filter == null)
+            {
+                filter = (IHubFilter)ActivatorUtilities.CreateInstance(serviceProvider, filterType);
+            }
```

Comment:
This could just call ActivatorUtilities.GetServiceOrCreateInstance.

The better approach is probably to continue calling GetService, and then call ActivatorUtilities.CreateFactory and store the result in a `ConcurrentDictionary<Type, ObjectFactory>` for types that need to be activated.

Reply:
You dont need a dictionary, just store the objectfactory as a field once it's resolved lazily.

Reply:
Actually it doesn't *need* to be lazy but whatever.

Reply:
> You dont need a dictionary, just store the objectfactory as a field once it's resolved lazily.

Good point. I didn't notice HubFilterFactory is always constructed in a generic method and that GetFilter is always called with the same type argument. Making this a `HubFilterFactory<T>` and storing the ObjectFactory directly in a field is way cleaner.